ARCHITECTURE.md

```diff
@@ -0,0 +1,456 @@
+# Architecture
+
+This document describes the high-level architecture of PyMC3.
+If you want to familiarize yourself with the code base, you are just in the right place!
+
+
+# Bird's Eye View
+
+TODO: Update diagram to include SMC
+![Architecture](docs/Architecture.png)
```

Comment:
I would try to show the most important dependencies in the image. Maybe something like distributions, gp and ode are used to "build" models, then samplers take a model and a step method to generate samples (and store these samples in a backend which seems to be missing but I'm not sure if its missing or backends are gone in v4). More nuance can then be added in the text and in the developer guide. 

Working on this more hierarchical image (like working on this doc) can help better define what and why should be a module/submodule, even how to structure the API doc page.

Reply:
The image could even be graphviz generated with each node being a link to the right section/respective api page if someone is up for the task.

References: https://www.sphinx-doc.org/en/master/usage/extensions/graphviz.html and https://jhermann.github.io/blog/python/documentation/2020/03/25/sphinx_ext_graphviz.html